File: 백준/Silver/4963. 섬의 개수/섬의 개수.cs
using System;
using System.Text;
using System.Collections.Generic;

class Program
{
    static int[] dx = new int[] {0,0,1,-1,1,-1,1,-1};
    static int[] dy = new int[] {1,-1,0,0,-1,1,1,-1};
    static int mx,my,land;
    
    static bool[,] visited = new bool[51,51];
    static int[,] map = new int[51,51];
    
    static void dfs(int x,int y) 
    {
        visited[x,y] = true;
        
        for(int i=0;i<8;i++)
        {
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if(nx >= 0 && ny >= 0 && nx < mx && ny < my)
        {
            if(map[nx,ny] == 1 && !visited[nx,ny]) dfs(nx,ny);
        }
        }
    }
    
    static void Main() {
        
        while(true)
        {
            land = 0;
            string[] str = Console.ReadLine().Split(' ');
            my = int.Parse(str[0]);
            mx = int.Parse(str[1]);
            if(mx == 0 && my == 0) break;
            for(int i=0;i<mx;i++)
            {
                string[] input = Console.ReadLine().Split(' ');
                
                for(int j=0;j<my;j++)
                {
                    map[i,j] = int.Parse(input[j]);
                    visited[i,j] = false;
                }
            }
            for(int i=0;i<mx;i++)
            {
                for(int j=0;j<my;j++)
                {
                    if(map[i,j] == 1 && !visited[i,j])
                    {
                        land++;
                        dfs(i,j);
                    }
                }
            }
            Console.WriteLine(land);
        }
    }
}

File: 백준/Silver/4963. 섬의 개수/섬의 개수.py
import sys
sys.setrecursionlimit(10**6)

dx = [0,0,1,-1,1,-1,1,-1]
dy = [1,-1,0,0,-1,1,1,-1]

def dfs(x,y):
    visited[x][y] = True
    for i in range(0,8):
        nx = x + dx[i]
        ny = y + dy[i]
        if nx >= 0 and ny >= 0 and nx < mx and ny < my:
            if edge[nx][ny] == 1 and not visited[nx][ny]: dfs(nx,ny)
            
while True:
    land = 0
    my,mx = map(int,input().split())
    if mx==0 and my==0: break

    edge = []
    visited = [[False] * 51 for _ in range(51)]
    
    for i in range(0,mx):
        edge.append(list(map(int,input().split())))
        for j in range(0,my):
            visited[i][j] = False
   
    for i in range(0,mx):
        for j in range(0,my):
            if edge[i][j] == 1 and not visited[i][j]:
                land+=1
                dfs(i,j)
    print(land)